Truncate Launchpad description text

jtarricone opened the pull request:
This implements a simple truncation on the description elements in the launch pad. When the text elements are over a certain height, the element is resized, a '...' is shown at the end, and a toggle button is added.

![download 2](https://cloud.githubusercontent.com/assets/18290666/21155055/a7a19e6c-c13e-11e6-868f-791ccac4acea.png)
![download 3](https://cloud.githubusercontent.com/assets/18290666/21155054/a7a1af2e-c13e-11e6-954e-2fff88c3a647.png)


To test:
* clone this branch, bring up the environment, and open the app
* click on the sidebar item to open the launchpad
* check that the first item is truncated and has a "View more" button
* check that clicking the button toggle the text truncation
* check that the other items are not truncated

Connects #749

jtarricone commented:
Ignore all but the last commit. I'll rebase this as soon as #763 is merged.

kdeloach commented:
Mind removing these unrelated commits? Then you won't have to rebase when other pull requests are merged.

jtarricone commented:
I rebased on `prototype`, this PR is independent of my other ones.

mmcfarland commented:
I believe the referenced issue describes the text associated with a plugin, which are the elements under the"start exploring" header and not the main text body.  It may be useful to have it there as well, but it needs to be for the plugins primarily.

jtarricone commented:
I'll update this to remove the truncation of the main text, but as it stands the plugin descriptions are also truncated the same way, but none them are currently long enough trigger the adjustment. What is the height limit for those?

mmcfarland commented:
The original request wasn't expressed as a height limit but rather, to be "twitter-like", which is to say a few sentences.  I'm not sure how that translates to height, or if things like dpi would affect that across different computers.

kdeloach commented on src/GeositeFramework/plugins/launchpad/main.js:
```diff
@@ -86,6 +86,33 @@ define([
+                        });
+                    } else {
+                        $(".truncate-toggle", this).css({ display:"none" });
+                    }
```
Couple of suggestions:

* This code does exactly what #749 suggests ("If the plugin description is greater than a given height ..."). However, I would recommend using text length instead of DOM height. Then we could move half of this logic to the template.

  Ex.

  ```
  <p><%= shortDescription %></p>
  <% if (description.length > 100) { %>
  <button class="button truncate-toggle">View more</button>
  <% } %>
  ```

* Consider moving the click events to the `bindEvents` function. We try to use live selectors whenever possible.
* For parity with the rest of the codebase, prefer to use jQuery functions to manipulate DOM attributes (ex. `$(this).text(...)` instead of `this.textContent`, and `$(this).toggle(...)` instead of `this.classList.toggle`)
* Defining hard-coded strings in code instead of templates presents some internationalization difficulties. Consider using an ellipsis instead to avoid this problem? Maybe even get rid of the option to "View Less" to keep things simple.

jtarricone replied:
* Fair enough, but basing the truncation on element height rather than text length has the advantage of maintaining consistent behavior regardless of what the font size is or how it's defined (if that's the desired behavior). I don't think having the condition in the template is practically much simpler than evaluating it at runtime, respective of each criterion, but ultimately there should be some clarification of what the original request was
* Good point, I'll move the block to `bindEvents`
* Good point about using jQuery functions, I'll update those
* I used the "view more" label because that seemed to be implied in #749. I would think an ellipsis and/or up/down icons would be enough to get the intent across. Whether nixing the "view less" option is a good idea rather depends on how long the text blocks are expected to be.

kdeloach replied:
I agree this is more of a design issue. But I would argue that using text length is simpler and looks better. Making sure the "truncate container" height lines up with the text will be an extra chore.

See the difference between these two options:

1. Height
  ![image](https://cloud.githubusercontent.com/assets/43062/21188046/0146628c-c1e8-11e6-96a9-f88eb7960cc3.png)

2. Text length
  ![image](https://cloud.githubusercontent.com/assets/43062/21188053/06828924-c1e8-11e6-94c9-8cb5f4c27427.png)

kdeloach commented on src/GeositeFramework/plugins/launchpad/style.css:
```diff
@@ -26,3 +26,41 @@
+    padding: 2px !important;
+    font-size: 0.7em !important;
+    width: 35%;
+    float: none !important;
```
I think `!important` should be reserved for extreme cases. Like when there is third party CSS we physically cannot modify. Using it everywhere makes styles harder to override (for example, in print mode).

jtarricone replied:
The basic functionality is unchanged if the `button` class is removed, but all the accumulated styling goes with it, so it seems using `!important` is probably the easiest way to go.

kdeloach replied:
In this case, you could increase the specificity of the rule to `.launchpad-plugin .plugins .button.truncate-toggle` to avoid using `!important`.

jtarricone commented:
The description in #749 mentioned height as the criterion, but regardless I think basing it on height is probably the most straightforward and possibly the most robust solution. I think `px` values, as far as CSS is concerned, are angular units as opposed to linear ones, so I think a height specified in `px` would be stable across viewers. There should maybe be some coupling between whatever the truncated height is and the font size within the container, once the desired effect is decided.

How long of text blocks are these intended to be?

jtarricone commented:
I just pushed an update that addresses the points that didn't require further clarification.

kdeloach commented on src/GeositeFramework/plugins/launchpad/main.js:
```diff
@@ -54,6 +54,19 @@ define([
+                        var parentEl = $(e.target.parentElement);
+                        $('p', parentEl).each(function() {
+                            $(this).toggleClass('truncate-box');
+                        });
```
I think the jQuery way of doing this would be something like `$el.find('p').toggleClass(...)`.

kdeloach commented on src/GeositeFramework/plugins/launchpad/main.js:
```diff
@@ -54,6 +54,19 @@ define([
+                                $(this).text('View less');
+                            } else {
+                                $(this).text('View more');
+                            }
```
We should handle i18n before merging.

jtarricone replied:
I'm not sure how literal the blurb in #749 was intended to be, but again I'd think and ellipsis and/or arrow icons would be sufficient.

kdeloach commented on src/GeositeFramework/plugins/launchpad/main.js:
```diff
@@ -86,6 +99,18 @@ define([
+                            $(this).toggleClass('truncate-box');
+                        });
+                    } else {
+                        $(".truncate-toggle", this).css({ display:"none" });
```
Consider using jQuery's `hide` and `show` functions instead.

ajrobbins commented:
Chiming in here:
- Original "height" request came from the client and should not be taken literally if there is a better way to implement the limitation across browsers.
- The idea of "Twitter-length" was suggested, so if we could limit to 140 characters (and that length looks good in this layout) that would be one option
- Needs to work with i18n
- I like the ellipses and "View More". Tagging @jfrankl who will likely follow with or have style recommendations

jtarricone commented:
After the clarifications, I updated this to truncate at 140 characters and refactored where necessary. I updated the `region.json` file to contain a description longer than 140 characters just for purposes of always having this feature on the launchpad.

jfrankl commented:
I would recommend implementing a 140 character maximum on plugin descriptions rather than use a view more text function like this. I think we should be encouraging plugin developers to write brief, easy to understand descriptions. Have people run into problems where they need to write longer descriptions, or is this predicting a future issue?

jfrankl commented:
I don't think the Launchpad is a good place for a "long" or "detailed" description. It should contain a straightforward introduction. If more of an explanation is needed, could it be in the application itself?

jtarricone commented:
I pushed another update that implements the i18n text for the control

ajrobbins commented:
@jfrankl I agree with you in principle, but based on specific requests from TNC I think we need to allow for the possibility that someone writes a description that's longer than 140 characters.

jtarricone commented:
What's the verdict on this?

mmcfarland commented:
I'd recommend text length restriction over height.

jtarricone commented:
@mmcfarland it's been updated to truncate at 140 characters. I meant verdict with respect to whether it needs to be merged, or if the feature is no longer required.

mmcfarland commented:
Still required.

kdeloach commented on src/GeositeFramework/plugins/launchpad/style.css:
```diff
@@ -26,3 +26,41 @@
+
+.launchpad-plugin .plugins .button.truncate-toggle, .scenarios .button.truncate-toggle {
+    padding: 2px !important;
+    font-size: 0.7em !important;
```
Consider removing `!important` from these lines. There didn't seem to be a visual difference when I removed them during testing.

kdeloach commented on src/GeositeFramework/plugins/launchpad/style.css:
```diff
@@ -26,3 +26,41 @@
+    overflow: hidden;
+    position: relative;
+    line-height: 1.2em;
+    max-height: 2.6em;
```
I don't think we need to clip this container by height anymore, now that we're truncating the text to 140 characters. We may not need this container at all.

jtarricone replied:
The text itself isn't truncated, the box class is conditioned on the text length. The visible height is the mechanism of hiding the text.

kdeloach replied:
Let's discuss in person.

kdeloach replied:
Might make more sense to address this in a future PR if it becomes an issue.

kdeloach commented on src/GeositeFramework/plugins/launchpad/style.css:
```diff
@@ -26,3 +26,41 @@
+    bottom: 2px;
+}
+
+.truncate-box:after {
```
I removed this `.truncate-box:after` block during testing and it made no visual difference. Consider removing.

kdeloach commented on src/GeositeFramework/plugins/launchpad/style.css:
```diff
@@ -26,3 +26,41 @@
+}
+
+.truncate-container > p {
+    width: 90%;
```
The text width is much narrower than the original design. I think some of this CSS and markup changes should be changed to preserve the original style.

Before:
![image](https://cloud.githubusercontent.com/assets/43062/21296958/13022186-c545-11e6-924f-f6b23704b2c5.png)

After:
![image](https://cloud.githubusercontent.com/assets/43062/21296946/eee8cb60-c544-11e6-8486-805ed02f8292.png)

kdeloach commented on src/GeositeFramework/plugins/launchpad/main.js:
```diff
@@ -54,6 +54,18 @@ define([
+                            } else {
+                                $(this).attr('data-i18n', 'See More');
+                            }
+                            $(this).localize();
```
Imagine how one might implement this feature using React. Clicking on the toggle button would update the component state, then we would re-render the component using the new state. 

That is the general design pattern that we are trying to follow with TNC plugins, and that's what I would recommend that we do here. The result should be a clean separation between the state and the template. No DOM manipulation should be necessary.

kdeloach replied:
Ultimately, we shouldn't be making decisions based on i18n attributes. We can pair on this.

kdeloach commented on src/GeositeFramework/plugins/launchpad/templates.html:
```diff
@@ -23,7 +28,12 @@
+                        <div class="truncate-container">
+                            <p><%= scenario.description %></p>
+                            <% if (scenario.description.length > 140) { %>
+                                <button class="button truncate-toggle i18n" data-i18n="[title]See Less" title="See Less"></button>
```
Not sure if this was intentional, but the button markup is a little different for scenarios compared to plugins. I noticed that the tooltip doesn't always match the button text.

![image](https://cloud.githubusercontent.com/assets/43062/21297159/9ae8e7b2-c548-11e6-8f27-cced0960dad4.png)

jtarricone commented on src/GeositeFramework/plugins/launchpad/main.js:
```diff
@@ -86,6 +98,7 @@ define([
                 }));
 
                 $(this.container).empty().append($el);
+                $('.truncate-toggle').click();
```
It can be done in the template instead.

kdeloach commented on src/GeositeFramework/locales/en.json:
```diff
@@ -69,4 +69,6 @@
     "Length:": "Length:",
     "Double-click to end a line segment. Single-click the first node to end a polygon.": "Double-click to end a line segment. Single-click the first node to end a polygon.",
     "Click the map to start measuring": "Click the map to start measuring"
+    "See More": "See More",
```
We need trailing commas on line 71 to make this valid JSON.

kdeloach commented on src/GeositeFramework/plugins/launchpad/style.css:
```diff
@@ -26,3 +26,28 @@
     padding: 0;
 }
+
+.launchpad-plugin .plugins .button.truncate-toggle, .scenarios .button.truncate-toggle {
```
Minor, but consider adding a line break after the comma so it's easier to tell this style has 2 associated rules.

kdeloach commented on src/GeositeFramework/plugins/launchpad/templates.html:
```diff
@@ -23,7 +30,14 @@
+                        <div class="truncate-container">
+                            <% if (scenario.description.length > 140) { %>
+                                <p class="truncate-box"><%= scenario.description %></p>
+                                <button class="button truncate-toggle i18n" data-i18n="[title]See More" title="See More"></button>
```
Not sure if this was intentional, but scenario buttons have a title attribute, but the plugin buttons do not.

jtarricone commented:
After clarifying some points, I refactored this a bit to instead use a state object to avoid doing direct DOM manipulation.
